azurelinuxagent/ga/firewall_manager.py

```diff
@@ -0,0 +1,499 @@
+
+    def _get_drop_rule_command(self, command_option):
+        return self._base_command + [command_option, "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "DROP"]
+
+    def _get_legacy_rule_command(self, command_option):
+        # There was a rule change at 2.2.26, which started dropping non-root traffic to WireServer. The previous rule allowed traffic, and needs to be removed
+        # for the newer DROP rule to have any effect. This function returns the command to manipulate the legacy rule that was added <= 2.2.25. Until 2.2.25
+        # has aged out, keep this cleanup in place.
+        return self._base_command + [command_option, "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "ACCEPT"]
+
+    def _get_append_command_option(self):
+        return "-A"
+
+    def _get_check_command_option(self):
+        return "-C"
+
+    def _get_delete_command_option(self):
+        return "-D"
+
+
+class FirewallCmd(_FirewallManagerMultipleRules):
+    """
+    FirewallManager based on the firewalld command-line tool.
+    """
+    def _get_state_command(self):
+        return ["firewall-cmd", "--permanent", "--direct", "--get-all-passthroughs"]
+
+    def _get_accept_dns_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", '--destination-port', '53', '-j', 'ACCEPT']
+
+    def _get_accept_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "owner", "--uid-owner", str(os.getuid()), "-j", "ACCEPT"]
+
+    def _get_drop_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "DROP"]
+
+    def _get_legacy_rule_command(self, command_option):
+        # Agents <= 2.7.0.6 inserted (-I) the rule to accept DNS traffic; later agents changed that to append (-A) the rule.
+        # The insert rule needs to be removed, otherwise there will be duplicate rules for DNS.
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-I", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", '--destination-port', '53', '-j', 'ACCEPT']
+
+    def _get_append_command_option(self):
+        return "--passthrough"
+
+    def _get_check_command_option(self):
+        return "--query-passthrough"
+
+    def _get_delete_command_option(self):
+        return "--remove-passthrough"
+
+
+class NfTables(FirewallManager):
```

Comment:
Could you explain the distinction between FirewallManager and FirewallManagerMultipleRules? i want to understand why they have separate implementations

Reply:
This is my attempt to explain that in the current comments. 

`    FirewallManager abstracts the interface for managing the firewall rules on the WireServer address. Concrete implementations provide the underlying functionality using command-line tools that vary across distros (e.g. iptables, firewalld, and nftables.)
`

FirewallManagerMultipleRules
`
    Base class for firewall managers that handle multiple rules, each rule being manipulated with a different command line (e.g. iptables, firewalld)
 `
 
iptables and firewallcmd maintains 3 separate rules, that need to be manipulated with 3 separate command variations

nft maintains 1 single rule, with 1 single command